testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java

```diff
@@ -0,0 +1,630 @@
+
+        createClientSession(session, realm.getClientByClientId("test-app"), sessions[0], "http://redirect", "state");
+        createClientSession(session, realm.getClientByClientId("third-party"), sessions[0], "http://redirect", "state");
+
+        sessions[1] = session.sessions().createUserSession(realm, session.users().getUserByUsername("user1", realm), "user1", "127.0.0.2", "form", true, null, null);
+        createClientSession(session, realm.getClientByClientId("test-app"), sessions[1], "http://redirect", "state");
+
+        sessions[2] = session.sessions().createUserSession(realm, session.users().getUserByUsername("user2", realm), "user2", "127.0.0.3", "form", true, null, null);
+        createClientSession(session, realm.getClientByClientId("test-app"), sessions[2], "http://redirect", "state");
+
+        return sessions;
+    }
+
+    public static void reloadState(KeycloakSession session) {
+        reloadState(session, false);
+    }
+
+    public static void reloadState(KeycloakSession session, Boolean initialConfig) {
+        currentSession = session;
+        realm = currentSession.realms().getRealm("test");
+        if (initialConfig == true) {
+            currentSession.users().addUser(realm, "user1").setEmail("user1@localhost");
+            currentSession.users().addUser(realm, "user2").setEmail("user2@localhost");
+        }
+        sessionManager = new UserSessionManager(currentSession);
+        persister = currentSession.getProvider(UserSessionPersisterProvider.class);
+    }
+
+    @Before
+    public void before() {
+        testingClient.server().run(session -> {
+            KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionBefore) -> {
+                reloadState(sessionBefore, true);
+            });
+        });
+    }
+
+    @After
+    public void after() {
+        testingClient.server().run(session -> {
+            RealmModel realm = session.realms().getRealmByName("test");
+            session.sessions().removeUserSessions(realm);
+            UserModel user1 = session.users().getUserByUsername("user1", realm);
+            UserModel user2 = session.users().getUserByUsername("user2", realm);
+
+            UserManager um = new UserManager(session);
+            if (user1 != null) {
+                um.removeUser(realm, user1);
+            }
+            if (user2 != null) {
+                um.removeUser(realm, user2);
+            }
+        });
+    }
+
+    @Test
+    @ModelTest
+    public void testOfflineSessionsCrud(KeycloakSession session) {
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
+            // Create some online sessions in infinispan
+            int started = Time.currentTime();
+
+            reloadState(sessionCrud);
+
+            UserSessionModel[] origSessions = createSessions(currentSession);
+            sessionCrud.getTransactionManager().commit();
```

Comment:
As mentioned in the email, these calls to "sessionCrud.getTransactionManager().commit();" should be avoided. And nested transactions should be avoided too.

It's highly possible that even if nested transactions are passing with H2, they will probably fail with MySQL (and some other DBs).

In shortcut, if the old code contained something like this:
```
someMethodCalls1();

resetSession();

someMethodCalls2();
````

You can replace in the new with something like:

````
KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
                someMethodCalls1();
);

KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
                someMethodCalls2();
);
````
However not nested transactions.

Reply:
This applies for many places in this class